### DISCUSSION TITLE
How to get fitted learner from Object

### DISCUSSION THREAD

**Discussion by chungdz** [DoubleML/doubleml-for-py#217] [Q&A] [answered]
I try to get the fitted learner from DoubleMLPLR object.

I try learner/models attribute of this object. The former one gives me learners that haven't been fitted, the latter gives me None.

Any solutions please?

**Comment by SvenKlaassen** [comment]
Thanks for your question.
Could you please post an minimal reproducable example, such that I can replicate the error.

**Comment by chungdz** [comment]
Copy the example code in document with the error I show:

```
import numpy as np
from doubleml.datasets import fetch_bonus
from sklearn.base import clone
from sklearn.ensemble import RandomForestRegressor
from doubleml import DoubleMLData, DoubleMLPLR

df_bonus = fetch_bonus('DataFrame')
xinputs = ['female', 'black', 'othrace', 'dep1', 'dep2',
        'q2', 'q3', 'q4', 'q5', 'q6', 'agelt35', 'agegt54',
        'durable', 'lusd', 'husd']
dml_data_bonus = DoubleMLData(df_bonus,
                             y_col='inuidur1',
                             d_cols='tg',
                             x_cols=xinputs)

learner = RandomForestRegressor(n_estimators=500, max_features='sqrt', max_depth= 5)
ml_l_bonus = clone(learner)
ml_m_bonus = clone(learner)

obj_dml_plr_bonus = DoubleMLPLR(dml_data_bonus, ml_l_bonus, ml_m_bonus)
obj_dml_plr_bonus.fit()

obj_dml_plr_bonus.learner['ml_l'].predict(df_bonus[xinputs])

obj_dml_plr_bonus.models.predict(df_bonus[xinputs])
```

Error messages:
![image](https://github.com/DoubleML/doubleml-for-py/assets/24776624/dbb8297b-c4f0-43a4-b3ae-ae7d326c6b42)
![image](https://github.com/DoubleML/doubleml-for-py/assets/24776624/02189ea9-ec2b-425c-9d74-ab090b4624d7)

**Reply by SvenKlaassen** [reply, marked as answer]
Thank you. 
The `learner` attribute is just a copy of the unfitted learner (since they have to be used multiple times). 
To store the fitted models you have set the `store_models` argement in the `.fit()` method. The fitted models are then accessible under the `models` attribute.

```
import numpy as np
from doubleml.datasets import fetch_bonus
from sklearn.base import clone
from sklearn.ensemble import RandomForestRegressor
from doubleml import DoubleMLData, DoubleMLPLR

df_bonus = fetch_bonus('DataFrame')
xinputs = ['female', 'black', 'othrace', 'dep1', 'dep2',
        'q2', 'q3', 'q4', 'q5', 'q6', 'agelt35', 'agegt54',
        'durable', 'lusd', 'husd']
dml_data_bonus = DoubleMLData(df_bonus,
                             y_col='inuidur1',
                             d_cols='tg',
                             x_cols=xinputs)

learner = RandomForestRegressor(n_estimators=500, max_features='sqrt', max_depth= 5)
ml_l_bonus = clone(learner)
ml_m_bonus = clone(learner)

obj_dml_plr_bonus = DoubleMLPLR(dml_data_bonus, ml_l_bonus, ml_m_bonus)
obj_dml_plr_bonus.fit(store_models=True)

obj_dml_plr_bonus.models
```

This should look like this

```
{'ml_l': {'tg': [[RandomForestRegressor(max_depth=5, max_features='sqrt', n_estimators=500),
    RandomForestRegressor(max_depth=5, max_features='sqrt', n_estimators=500),
    RandomForestRegressor(max_depth=5, max_features='sqrt', n_estimators=500),
    RandomForestRegressor(max_depth=5, max_features='sqrt', n_estimators=500),
    RandomForestRegressor(max_depth=5, max_features='sqrt', n_estimators=500)]]},
 'ml_m': {'tg': [[RandomForestRegressor(max_depth=5, max_features='sqrt', n_estimators=500),
    RandomForestRegressor(max_depth=5, max_features='sqrt', n_estimators=500),
    RandomForestRegressor(max_depth=5, max_features='sqrt', n_estimators=500),
    RandomForestRegressor(max_depth=5, max_features='sqrt', n_estimators=500),
    RandomForestRegressor(max_depth=5, max_features='sqrt', n_estimators=500)]]}}
```

Remark that this are a couple of models which you can access via a nested dictionary where the first level is the nuisance element and the second element is the treatment name (potentially multiple treatments). This then includes a nested list as the first index refers to the `n_rep` argument (if the cross-fitting has been repeated multiple times [default is `1`]) and lastly to the `n_folds` argument (as there is one model for each fold [default is `5`]).
To access a single model you can use

```
obj_dml_plr_bonus.models['ml_l']['tg'][0][0].predict(df_bonus[xinputs])
```
I hope this works for you.

**Reply by kartechbabu** [reply]
Dear Sven,

Is it possible to get CI or SE for the predictions? 

Thanks
Kartik

**Reply by SvenKlaassen** [reply]
Hello @kartechbabu,

generally it is not possible to get the CI oder SE for these predictions from our package. For the PLR model one could in principle generate outcome predictions via

$$
\hat{y}:=\hat{\theta}D + \hat{g}(X)
$$

but this would include the noisy estimate of $\hat{g}(X)$.

**Reply by kartechbabu** [reply]
Thank you so much, @SvenKlaassen, for your prompt response!
